spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,197 @@
+
+  private final SparkSession spark;
+  private final JavaSparkContext sparkContext;
+  private final SerializableConfiguration hadoopConf;
+  private final FileIO fileIO;
+  private final int partitionDiscoveryParallelism;
+  private final String dataLocation;
+
+  private String allDataFilesTable = null;
+  private Long olderThanTimestamp = null;
+  private boolean dryRun = false;
+
+  RemoveOrphanFilesAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.hadoopConf = new SerializableConfiguration(spark.sessionState().newHadoopConf());
+    this.fileIO = table.io();
+    this.partitionDiscoveryParallelism = spark.sessionState().conf().parallelPartitionDiscoveryParallelism();
+    this.dataLocation = table.locationProvider().dataLocation();
+  }
+
+  public RemoveOrphanFilesAction allDataFilesTable(String newAllDataFilesTable) {
+    this.allDataFilesTable = newAllDataFilesTable;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction olderThan(long newOlderThanTimestamp) {
+    this.olderThanTimestamp = newOlderThanTimestamp;
+    return this;
+  }
+
+  public RemoveOrphanFilesAction dryRun(boolean newDryRun) {
+    this.dryRun = newDryRun;
+    return this;
+  }
+
+  @Override
+  public RemoveOrphanFilesActionResult execute() {
+    Preconditions.checkArgument(allDataFilesTable != null, "allDataFilesTable must be set");
+    Preconditions.checkArgument(olderThanTimestamp != null, "olderThanTimestamp should be set");
+
+    Dataset<Row> validDataFileDF = buildValidDataFileDF();
+    Dataset<Row> actualDataFileDF = buildActualDataFileDF();
+
+    Column joinCond = validDataFileDF.col("file_path").equalTo(actualDataFileDF.col("file_path"));
+    List<String> orphanDataFiles = actualDataFileDF.join(validDataFileDF, joinCond, "leftanti")
+        .as(Encoders.STRING())
+        .collectAsList();
+
+    if (!dryRun) {
+      Tasks.foreach(orphanDataFiles)
+          .noRetry()
+          .suppressFailureWhenFinished()
+          .onFailure((file, exc) -> LOG.warn("Failed to delete data file: {}", file, exc))
+          .run(fileIO::deleteFile);
+    }
+
+    return new RemoveOrphanFilesActionResult(orphanDataFiles);
+  }
+
+  private Dataset<Row> buildValidDataFileDF() {
+    return spark.read().format("iceberg")
+        .load(allDataFilesTable)
+        .select("file_path")
+        .distinct();
+  }
+
+  private Dataset<Row> buildActualDataFileDF() {
+    List<String> topLevelDirs = Lists.newArrayList();
+    List<String> matchingTopLevelFiles = Lists.newArrayList();
+
+    try {
+      Path dataPath = new Path(dataLocation);
+      FileSystem fs = dataPath.getFileSystem(hadoopConf.value());
+
+      for (FileStatus file : fs.listStatus(dataPath, HiddenPathFilter.get())) {
+        // TODO: handle custom metadata folders
+        // we need to ignore the metadata folder when data is written to the root table location
```

Comment:
Do we?

Reply:
It is a leftanti join between all files and valid data files. If we don't ignore the metadata folder, we will remove all metadata files, right?

Reply:
I think we want better protection for the metadata files. We could have a case where the user put metadata files into the data folder and then relocated the config to a different place. We still want to protect the metadata.

I think that means that we should do the anti-join with all data and metadata file paths. We can get recent metadata files from:
* The `all_manifests` table has manifest locations
* The `snapshots` table has manifest list locations
* The table metadata has the current metadata location and a list of old metadata locations

I think it should be sufficient to protect all of those paths. Then we don't need to worry about a prefix to protect metadata.

Reply:
Since we limit the number of previous metadata locations we keep in each new metadata file, the list of old metadata locations might not be complete in the current metadata file. It seems safe to remove others (assuming they are old enough). What do you think, @rdblue?

Reply:
I think it's a good thing to remove them.

Reply:
Added a test for old metadata files.